File: Solution O(1) space.js

/**
 * @param {number[]} input
 * @return {number[]}
 */
var findErrorNums = function (input) {
    const NOT_FOUND = -1;
    let duplicate = NOT_FOUND;
    let missing = NOT_FOUND;

    for (let i = 0; i < input.length; ++i) {
        if (input[Math.abs(input[i]) - 1] < 0) {
            duplicate = Math.abs(input[i]);
            continue;
        }
        input[Math.abs(input[i]) - 1] = -input[Math.abs(input[i]) - 1];
    }

    for (let i = 0; i < input.length && missing === NOT_FOUND; ++i) {
        if (input[i] > 0) {
            missing = i + 1;
        }
    }
    return [duplicate, missing];
};

File: Solution O(1) space.cpp

#include <vector>
using namespace std;

class Solution {
    
    inline static const int NOT_FOUND = -1;

public:
    vector<int> findErrorNums(vector<int>& input) const {
        int duplicate = NOT_FOUND;
        int missing = NOT_FOUND;

        for (int i = 0; i < input.size(); ++i) {
            if (input[abs(input[i]) - 1] < 0) {
                duplicate = abs(input[i]);
                continue;
            }
            input[abs(input[i]) - 1] = -input[abs(input[i]) - 1];
        }

        for (int i = 0; i < input.size() && missing == NOT_FOUND; ++i) {
            if (input[i] > 0) {
                missing = i + 1;
            }
        }
        return vector<int>{duplicate, missing};
    }
};
